src/librustc_metadata/decoder.rs

```diff
@@ -496,30 +467,106 @@ impl<'a, 'tcx> CrateMetadata {
+                .as_symbol()
+        } else {
+            self.get_proc_macros(sess).unwrap()[&item_index].name
+        }
     }
 
-    pub fn def_kind(&self, index: DefIndex) -> Option<DefKind> {
-        if !self.is_proc_macro(index) {
+    pub fn def_kind(&self, index: DefIndex, sess: &Session) -> Option<DefKind> {
+        if !self.is_proc_macro(index, sess) {
             self.entry(index).kind.def_kind()
         } else {
             Some(DefKind::Macro(
-                self.proc_macros.as_ref().unwrap()[index.to_proc_macro_index()].1.macro_kind()
+                self.get_proc_macros(sess).unwrap()[&index].ext.macro_kind()
             ))
         }
     }
 
     pub fn get_span(&self, index: DefIndex, sess: &Session) -> Span {
-        match self.is_proc_macro(index) {
-            true => DUMMY_SP,
-            false => self.entry(index).span.decode((self, sess)),
+        self.entry(index).span.decode((self, sess))
+    }
+
+
+    pub fn get_proc_macros(&self, sess: &Session) -> Option<FxHashMap<DefIndex, FullProcMacro>>{
+        let proc_macros = self.root.proc_macro_data.as_ref().map(|proc_macros| {
+            let proc_macros = proc_macros.decode((self, sess));
+            if sess.opts.debugging_opts.dual_proc_macros {
+                let host_lib = self.host_lib.as_ref().unwrap();
+                self.load_proc_macros(
+                    &host_lib.metadata.get_root(),
+                    self.span,
+                    proc_macros,
+                    sess
+                )
+            } else {
+                self.load_proc_macros(&self.root,self.span, proc_macros, sess)
+            }
+        });
+        proc_macros
+    }
+
+    /// Loads custom derive macros.
+    ///
+    /// Note that this is intentionally similar to how we load plugins today,
+    /// but also intentionally separate. Plugins are likely always going to be
+    /// implemented as dynamic libraries, but we have a possible future where
+    /// custom derive (and other macro-1.1 style features) are implemented via
+    /// executables and custom IPC.
+    fn load_proc_macros(&self, root: &CrateRoot<'_>,
+                          span: Span,
+                          macros: impl ExactSizeIterator<Item = ProcMacroData>,
+                          sess: &Session)
+                          -> FxHashMap<DefIndex, FullProcMacro> {
+
+
+        let decls = self.raw_proc_macros.unwrap();
+        if macros.len() != decls.len() {
+            span_bug!(span, "proc-macro data mismatch: metadata '{:?}', decls {:?}'",
+                      macros.len(), decls.len());
         }
+
+        let extensions = macros.zip(decls.iter()).map(|(proc_macro, &decl) | {
+            let (name, kind, helper_attrs) = match decl {
+                ProcMacro::CustomDerive { trait_name, attributes, client } => {
+                    let helper_attrs =
+                        attributes.iter().cloned().map(Symbol::intern).collect::<Vec<_>>();
```

Comment:
Would be a good idea to get these from the attributes and not encode them in the first place.

Reply:
I'm hesitant to touch the proc macro bridge code in this PR, given that I'm already modifying so many other parts of proc macro encoding/decoding. Would it make sense to put that in a separate PR?

Reply:
Okay so we can defer this.